be/src/exprs/agg/array_agg.h

```diff
@@ -3,88 +3,61 @@
 
     void update(FunctionContext* ctx, const Column** columns, AggDataPtr __restrict state,
                 size_t row_num) const override {
         const auto& column = down_cast<const InputColumnType&>(*columns[0]);
-        if constexpr (IsSlice<InputCppType>) {
-            this->data(state).update(ctx->impl()->mem_pool(), column.get_slice(row_num));
-        } else {
-            this->data(state).update(ctx->impl()->mem_pool(), column.get_data()[row_num]);
-        }
+        // TODO: update is random access, so we could not pre-reserve memory for State, which is the bottleneck
+        this->data(state).update(column, row_num, 1);
     }
 
     void process_null(FunctionContext* ctx, AggDataPtr __restrict state) const override {
-        this->data(state).null_count++;
+        this->data(state).append_null();
     }
 
     void merge(FunctionContext* ctx, const Column* column, AggDataPtr __restrict state, size_t row_num) const override {
+        // Array element is nullable, so we need to extract the data from nullable column first
         const auto* input_column = down_cast<const ArrayColumn*>(column);
-        auto datum_array = input_column->get(row_num).get_array();
-
-        for (size_t i = 0; i < datum_array.size(); i++) {
-            if (datum_array[i].is_null()) {
-                this->data(state).null_count++;
-            } else {
-                this->data(state).update(ctx->impl()->mem_pool(), datum_array[i].get<InputCppType>());
-            }
-        }
+        auto offset_size = input_column->get_element_offset_size(row_num);
+        auto& array_element = down_cast<const NullableColumn&>(input_column->elements());
+        auto* element_data_column = down_cast<const InputColumnType*>(ColumnHelper::get_data_column(&array_element));
+        size_t element_null_count = array_element.null_count(offset_size.first, offset_size.second);
+        DCHECK_LE(element_null_count, offset_size.second);
+
+        this->data(state).update(*element_data_column, offset_size.first, offset_size.second - element_null_count);
+        this->data(state).append_null(element_null_count);
     }
 
     void serialize_to_column(FunctionContext* ctx, ConstAggDataPtr __restrict state, Column* to) const override {
-        const auto& data = this->data(state).items;
-        const size_t null_count = this->data(state).null_count;
-
+        auto& state_impl = this->data(state);
         auto* column = down_cast<ArrayColumn*>(to);
-        auto& offsets = column->offsets_column()->get_data();
-        auto& elements_column = column->elements_column();
-
-        if (null_count > 0) {
-            elements_column->append_nulls(null_count);
-        }
-        for (size_t i = 0; i < data.size(); i++) {
-            elements_column->append_datum(data[i]);
-        }
-        offsets.emplace_back(offsets.back() + data.size() + null_count);
+        column->append_array_element(state_impl.data_column, state_impl.null_count);
```

Comment:
Does `deserialize` is supported for  ArrayAGG ?

Reply:
deserialize is done in merge method.  deserialize and merge at the same time.